### PR TITLE
Transaction synchronization

### LINKED ISSUE #173: Changes to withNewSession semantics make it useless for audit log listener
Per behavior changes in:
https://github.com/grails/grails-data-mapping/issues/1130

The call to `withNewSession` in the listener should be removed as it isn't occurring in the same transactional context as the caller anymore. 

Audit logging should be atomic to the transaction making the changes such that both commit or rollback together. Previously, `withNewSession` was useful to prevent contaminating the outer session.

### PR DESCRIPTION
**Pull request by felixscheinost**
Up to 4.x `AuditTrail` were written immediately e.g. when using `save(flush: true)` even when afterwards the transaction failed or was rolled back.

This PR is based on @longwa's work in `feature/4.0.0_wip` to queue the `AuditTrail`s and batch-commit them AFTER the surrounding transaction is comitted. 

We added further tests that revealed edge cases when using nested transactions.
Unfortunately we didn't find a way to fix those edge cases in a GORM agnostic way yet.

Additionally, this supports using `hibernate.allow_update_outside_transaction: true`. This is handy when upgrading an old Grails 3 application to Grails 4 where `hibernate.allow_update_outside_transaction` defaults to false. 

For now we added a `provided` dependency on `org.grails.plugins:hibernate5`. This must be moved to a separate dependency  to support e.g. using solely MongoDB. 

This fixes #173 and #175.

### CODE REVIEW

File: plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            // DomainInDatastoreOne.withNewTransaction {
+            //   DomainInDatastoreTwo.withNewTransaction {
+            //     <some operation on DomainInDatastoreOne> + flush
+            //   }
+            // }
+            // We are able to register a synchroniation on the OUTER transaction even though this code is running while the
+            // INNER transaction is active.
+            //
+            // Just using Spring Transactions and/or GORM would be better because it would allow to be datastore agnostic
+            // But simply using TransactionSynchronizationManager.registerSynchronization doesn't work because it would register the synchronization in the innermost transaction.
```
Comment (robertoschwald) [inline plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy:71]:
```suggestion
            // But simply using TransactionSynchronizationManager.registerSynchronization doesn't work because it would 
            // register the synchronisation in the innermost transaction.
```

File: plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            // Seems like no transaction is active
+            //  => Save audit entry right now
+            // In Hibernate > 5.2 this can only happen by setting `hibernate.allow_update_outside_transaction: true`
+            //
+            // When audit domain class is in same datastore as a newly INSERTed entity we can't cause a flush of the session here.
+            // This would cause a Hibernate AssertionFailure: "null id in test.Author entry (don't flush the Session after an exception occurs)"
+            //
+            //  => we can't cause a session flush of the session that is used to flush the changes to the observed entity
+            //    => we can't use withNewTransaction because it reuses the current session
+            //  => we should still use a transaction because in theory the audit domain could be in another datastore where allow_update_outside_transaction isn't set
```
Comment (robertoschwald) [inline plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy:43]:
```suggestion
            //  => we should still use a transaction because in theory the audit domain could be in another datastore
            // where allow_update_outside_transaction isn't set
```

Comment (robertoschwald) [inline plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy]:
```suggestion
            //  => we should still use a transaction because in theory the audit domain could be in another datastore 
            // where allow_update_outside_transaction isn't set
```

File: plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            //  => Save audit entry right now
+            // In Hibernate > 5.2 this can only happen by setting `hibernate.allow_update_outside_transaction: true`
+            //
+            // When audit domain class is in same datastore as a newly INSERTed entity we can't cause a flush of the session here.
+            // This would cause a Hibernate AssertionFailure: "null id in test.Author entry (don't flush the Session after an exception occurs)"
+            //
+            //  => we can't cause a session flush of the session that is used to flush the changes to the observed entity
+            //    => we can't use withNewTransaction because it reuses the current session
+            //  => we should still use a transaction because in theory the audit domain could be in another datastore
+            // where allow_update_outside_transaction isn't set
```
Comment (robertoschwald) [inline plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogQueueManager.groovy:44]:
```suggestion
            //  => we should still use a transaction because in theory the audit domain could be in another datastore
            //        where allow_update_outside_transaction isn't set
```
